sched/environ/env_dup.c

```diff
@@ -61,82 +61,98 @@
  *
  ****************************************************************************/
 
-int env_dup(FAR struct task_group_s *group)
+int env_dup(FAR struct task_group_s *group, FAR char * const *envp_in)
 {
-  FAR struct tcb_s *ptcb = this_task();
   FAR char **envp = NULL;
   size_t envc = 0;
   int ret = OK;
 
-  DEBUGASSERT(group != NULL && ptcb != NULL && ptcb->group != NULL);
+  DEBUGASSERT(group != NULL);
 
   /* Pre-emption must be disabled throughout the following because the
    * environment may be shared.
    */
 
   sched_lock();
 
-  /* Does the parent task have an environment? */
+  /* Is there an environment ? */
 
-  if (ptcb->group != NULL && ptcb->group->tg_envp != NULL)
+  if (envp_in == NULL)
     {
-      /* Yes.. The parent task has an environment allocation. */
+      /* No, does the parent task have an environment? */
+
+      FAR struct tcb_s *ptcb = this_task();
 
-      while (ptcb->group->tg_envp[envc] != NULL)
+      DEBUGASSERT(ptcb != NULL && ptcb->group != NULL);
+
+      if (ptcb->group->tg_envp == NULL)
         {
-          envc++;
+          /* No, get out */
+
+          goto out_noenv;
         }
 
-      /* A special case is that the parent has an "empty" environment
-       * allocation, i.e., there is an allocation in place but it
-       * contains no variable definitions and, hence, envc == 0.
-       */
+      /* Yes.. The parent task has an environment allocation. */
 
-      if (envc > 0)
-        {
-          /* There is an environment, duplicate it */
+      envp_in = (FAR char * const *)ptcb->group->tg_envp;
```

Comment:
Sorry, it's too late to access tg_envp here, so it's better to do in exec_module.
So, we simplify replace ptcb->group->tg_envp with envp_in.

Reply:
This works for kernel threads (copying between kernel <-> kernel) so I'd leave this here. I'll patch binfmt still

Reply:
But kernel thread should have the empty environment variables, like how we handle file handle for kernel thread: https://github.com/apache/incubator-nuttx/pull/5391

Reply:
The env variables are not empty, the initial environment is set up for the idle thread and *all* threads including kernel threads inherit this environment. I never quite understood what the kernel threads use the env for though.

I can make it so that the env is not duplicated for kernel threads if you want ?

Reply:
> The env variables are not empty, the initial environment is set up for the idle thread and _all_ threads including kernel threads inherit this environment. I never quite understood what the kernel threads use the env for though.

Basically, kernel thread forbid to use any userspace api: https://github.com/apache/incubator-nuttx/issues/1108

> I can make it so that the env is not duplicated for kernel threads if you want ?
> 
> EDIT: There is one exception unfortunately, the proxy spawn thread. If this does not have an environment, then the first user task will have an empty environment too.

I forget this usage case.

> 
> ```
> static inline void nx_start_application(void)
> {
> #ifdef CONFIG_INIT_ARGS
>   FAR char *const argv[] =
>   {
>     CONFIG_INIT_ARGS,
>     NULL,
>   };
> #else
>   FAR char *const *argv = NULL;
> #endif
> ...
> #ifndef CONFIG_ARCH_ADDRENV
>   attr.stacksize = CONFIG_INIT_STACKSIZE;
> #endif
>   ret = exec_spawn(CONFIG_INIT_FILEPATH, argv,
>                    CONFIG_INIT_SYMTAB, CONFIG_INIT_NEXPORTS, &attr);
>   DEBUGASSERT(ret >= 0);
> #endif
> ```
> 
> I don't know enough about the history and how this is supposed to work, I just know how it works now. One option maybe would be to do it like argv, make initial envp here instead of doing it for the idle thread ?

Let's keep your code as it. I will refine proxy thread(this and spawn) later.

> 
> Maybe this is not an issue, the proxy thread is not a kernel thread right ?

No, it's a kernel thread.